modules/packages/DistributedBag.chpl

```diff
@@ -64,631 +62,788 @@ use BlockDist;
+  */
+  config const distributedBagWorkStealingRatio = 0.25;
+  /*
+    The maximum amount of work to steal from a horizontal node's segment. This
+    should be set to a value, in megabytes, that determines the maximum amount of
+    data that should be sent in bulk at once. The maximum number of elements is
+    determined by: (:const:`distributedBagWorkStealingMemCap` * 1024 * 1024) / sizeof(:type:`eltType`).
+    For example, if we are storing 8-byte integers and have a 1MB limit, we would
+    have a maximum of 125,000 elements stolen at once.
+  */
+  config const distributedBagWorkStealingMemCap : real = 1.0;
+  /*
+    The minimum number of elements a horizontal segment must have to become eligible
+    to be stolen from. This may be useful if some segments produce less elements than
+    others and should not be stolen from.
+  */
+  config const distributedBagWorkStealingMinElems = 1;
+  /*
+    The maximum amount of elements in an unroll block. This is crucial to ensure memory
+    usage does not rapidly grow out of control.
+  */
+  config const distributedBagMaxBlockSize = 1024 * 1024;
+
+  /*
+    Reference counter for DistributedBag
+  */
+  pragma "no doc"
+  class DistributedBagRC {
+    type eltType;
+    var _pid : int;
+
+    proc deinit() {
+      coforall loc in Locales do on loc {
+          delete chpl_getPrivatizedCopy(DistributedBagImpl(eltType), _pid);
       }
     }
-
-    return sz.read();
   }
 
   /*
-    Performs a lookup to determine if the requested element exists in this bag.
-    This method is best-effort and can be non-deterministic for concurrent
-    updates across nodes, and may miss elements resulting from any concurrent
-    insertion or removal operations.
+    A parallel-safe distributed multiset implementation that scales in terms of
+    nodes, processors per node (PPN), and workload; The more PPN, the more segments
+    we allocate to increase raw parallelism, and the larger the workload the better
+    locality (see :const:`distributedBagInitialBlockSize`). This data structure is unordered and employs
+    its own work-stealing algorithm, and provides a means to obtain a privatized instance of
+    the data structure for maximized performance.
   */
-  proc contains(elt : eltType) : bool {
-    var foundElt : atomic bool;
-    forall elem in getPrivatizedThis {
-      if elem == elt {
-        foundElt.write(true);
+  record DistBag {
+    type eltType;
+
+    /*
+      The implementation of the Bag is forwarded. See :class:`DistributedBagImpl` for
+      documentation.
+    */
+    var _impl : DistributedBagImpl(eltType);
+
+    // Privatized id...
+    pragma "no doc"
+    var _pid : int = -1;
+
+    // Reference Counting...
+    pragma "no doc"
+    var _rc : Shared(DistributedBagRC(eltType));
+		
+		pragma "no doc"
+    proc DistBag(type eltType, targetLocales = Locales) {
+      _pid = (new DistributedBagImpl(eltType, targetLocales = targetLocales)).pid;

```

Comment:
I would have expected this constructor to initialize _impl as well

Reply:
`_impl` is ignored for now, I merely have it for documentation sake. We obtain the privatized copy each time in `_value` and use that. I needed it because `forwarding` shows weird unhelpful stuff in `chpldoc`

Reply:
I think in that case it'd be more reasonable to chpldoc _value but I can believe that the _impl version might look nicer. Anyway if that's the case you can add a `//` comment to say that this field exists only for documentation purposes.